### DISCUSSION TITLE
[resteasy-reactive] Custom query param key extractor

### DISCUSSION THREAD

**Discussion by sliw39** [quarkusio/quarkus#35773] [Q&A]
Hi all,

I'm trying to implement a RESTeasy endpoint that matches the HL7 FHIR specification for "search" (https://www.hl7.org/fhir/search.html#modifiers).

I'm struggling with the "modifier" aspect of the implementation. For example, this request 
`GET [base]/Patient?family:contains=son`

will fall into the following Java code:
```java
@GET
public List<Patient> searchPatient(@QueryParam("family:contains") String familly) {
   // code
} 
```

But I don't want to create a method for every parameters and every modifiers.

Is there a way to have an implementation like this ?
```java
@GET
public List<Patient> searchPatient(@CustomQueryParam("family") StringParam family) {
   String modifier = family.modifier;
   String value = family.value;
} 
```

Thanks for your help !

[a comment by an automated account is not shown]

**Comment by geoand** [comment]
Have you had a look at the [ParamConverter](https://jakarta.ee/specifications/restful-ws/2.1/apidocs/javax/ws/rs/ext/paramconverter) API?

**Reply by sliw39** [reply]
Yes I did, but as I understood, it works on the value part of the query param, not on the key.
Here in my example, the syntax is `key:modifier=value`,

**Reply by geoand** [reply]
Ah, I see. 

I don't have any good solution for you at this point.

**Comment by luca-bassoricci** [comment]
1. Manual parsing of query params or 
2. a custom paramconverter which (de)serialize an object holding all `modifiers` as a single query param?

**Reply by sliw39** [reply]
1. This is actualy a solution I consider, but I will not be able to generate my interfaces with openAPI anymore.
2. That could be also a better solution, but I must follow the FHIR spec, so my delimiter must be a ":", not a "."

Digging into the code, I ran over [QueryParamExtractor](https://github.com/quarkusio/quarkus/blob/main/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/parameters/QueryParamExtractor.java) which is used [here](https://github.com/quarkusio/quarkus/blob/61eb4190bd28ae9eb0a6480e327e75e2f3064f24/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/startup/RuntimeResourceDeployment.java#L696)

Maybe there is a way to overload or configure the delimiter ?

**Reply by luca-bassoricci** [reply]
1. OpenAPI can be easily manual described; of course can be a tedious task
2. Sorry, I was incorrect and I edited my post.

All above points requires you have to manual describe OpenAPI so manual parsing seems to be the better solution and give you a readable query string.

`QueryParamExtractor` seems to be a sort of glue code used to call the `real` parameter extractor using the right API detected during build phase of application; @geoand for sure can give the right explanation about what `QueryParamExtractor` is used for :)

**Comment by sliw39** [comment]
We found a workaround that may help someone one day : 
We declared a VertX `@RouteFilter(400)` to rewrite the incomming url to a format understood by Resteasy.
`key:modifier=value` is transformed into `key.modifier=value` then use a `QueryParamExtractor`

**Reply by geoand** [reply]
Good idea!
